### PR TITLE
Time stepping backwards in time

### LINKED ISSUE #2310: Time stepping backwards in time

### PR DESCRIPTION
**Pull request by HenKlei**
Resolves #2310.

Time stepping backwards is now possible by choosing a larger `initial_time` compared to the `end_time`.

### CODE REVIEW

**Review by sdrave — COMMENTED** [review]
Just wondering: Do we want this on the level of TimeSteppers? One could also consider having a `MuMappedOperator`, which applies some tranformation to `mu` before passing it to the wrapped operator (like mapping `t` to `T-t`). Then `compute` could wrap the `Operator` and `RHS` before passing it to the timestepper. (Or this could happen in the `TimeStepper` interface methods.) Not sure about the advantages/disadvantages.

**Comment by HenKlei** [comment]
> Just wondering: Do we want this on the level of TimeSteppers? One could also consider having a `MuMappedOperator`, which applies some tranformation to `mu` before passing it to the wrapped operator (like mapping `t` to `T-t`). Then `compute` could wrap the `Operator` and `RHS` before passing it to the timestepper. (Or this could happen in the `TimeStepper` interface methods.) Not sure about the advantages/disadvantages.

Could be an option as well. But would that actually be enough? Assume the operator is independent of time. If I understand your suggestion correctly, the backward time stepping would behave exactly like the forward time stepping, right? Or do I miss anything here?

**Comment by sdrave** [comment]
> > Just wondering: Do we want this on the level of TimeSteppers? One could also consider having a `MuMappedOperator`, which applies some tranformation to `mu` before passing it to the wrapped operator (like mapping `t` to `T-t`). Then `compute` could wrap the `Operator` and `RHS` before passing it to the timestepper. (Or this could happen in the `TimeStepper` interface methods.) Not sure about the advantages/disadvantages.
> 
> Could be an option as well. But would that actually be enough? Assume the operator is independent of time. If I understand your suggestion correctly, the backward time stepping would behave exactly like the forward time stepping, right? Or do I miss anything here?

I'm thinking about something like

```python
class TimeReversedOperator(Operator):
    def __init__(self, T, op):
        ...
    def apply(self, U, mu):
        mu = mu.with_(t = self.T - mu['t'])
        return self.op.apply(U, mu=mu)
```

Yes, if neither operator nor rhs would depend on time, this would be the same as forward time stepping. Should it be different?

**Comment by HenKlei** [comment]
@HenKlei make sure to finish this...

**Comment by HenKlei** [comment]
Ping @sdrave

File: src/pymor/algorithms/timestepping.py [hunk 1]
```diff
@@ -202,6 +204,10 @@ def iterate(self, initial_time, end_time, initial_data, operator, rhs=None, mass
         if mu is None:
             mu = Mu()
 
+        sign = 1
+        if initial_time > end_time:
+            sign = -1
```
Comment (sdrave) [inline src/pymor/algorithms/timestepping.py:209]:
```suggestion
        sign = np.sign(end_time - initial_time)
```

File: src/pymor/algorithms/timestepping.py [hunk 2]
```diff
@@ -269,19 +275,23 @@ def iterate(self, initial_time, end_time, initial_data, operator, rhs=None, mass
...
         t = t0
+        yield U0, t
+
         U = U0.copy()
         if mu is None:
             mu = Mu()
 
+        sign = 1
+        if initial_time > end_time:
+            sign = -1
```
Comment (sdrave) [inline src/pymor/algorithms/timestepping.py:287]:
```suggestion
        sign = np.sign(end_time - initial_time)
```

File: src/pymor/algorithms/timestepping.py [hunk 3]
```diff
@@ -380,6 +390,10 @@ def iterate(self, initial_time, end_time, initial_data, operator, rhs=None, mass
...
```
Comment (sdrave) [inline src/pymor/algorithms/timestepping.py:395]:
```suggestion
        sign = np.sign(end_time - initial_time)
```

**Review by sdrave — CHANGES_REQUESTED** [review]
Looks good to me. I have only minor stylistic suggestions. What is missing though is some documentation that `final_time` can also be smaller than `initial_time` and what this means.

**Comment by HenKlei** [comment]
> Looks good to me. I have only minor stylistic suggestions. What is missing though is some documentation that `final_time` can also be smaller than `initial_time` and what this means.

Thanks for taking another look. I added a brief explanation, would that be enough?

**Comment by sdrave** [comment]
> > Looks good to me. I have only minor stylistic suggestions. What is missing though is some documentation that `final_time` can also be smaller than `initial_time` and what this means.
> 
> Thanks for taking another look. I added a brief explanation, would that be enough?

I'm not sure if it's mathematically clear what 'backwards time stepping' means. Maybe one should rather write that the corresponding terminal-value problem is solved?

**Comment by HenKlei** [comment]
> I'm not sure if it's mathematically clear what 'backwards time stepping' means. Maybe one should rather write that the corresponding terminal-value problem is solved?

You are right. I am still not sure if the new formulation is better.
And while changing the documentation, I was wondering whether one should actually change the order of the result. If one thinks of backwards time stepping the current order makes sense, however, if one sees it as a terminal value problem, the order should be reversed.

**Comment by sdrave** [comment]
> > I'm not sure if it's mathematically clear what 'backwards time stepping' means. Maybe one should rather write that the corresponding terminal-value problem is solved?
> 
> You are right. I am still not sure if the new formulation is better. And while changing the documentation, I was wondering whether one should actually change the order of the result. If one thinks of backwards time stepping the current order makes sense, however, if one sees it as a terminal value problem, the order should be reversed.

I like the 'terminal-value problem' formulation because at least it is a well-defined mathematical problem, and I think it makes more sense to speak about the problem to be solved than what a particular implementation might do.

Regarding the ordering: I guess the problem is that we use `initial_time` for the terminal value. If you think about it being the terminal value, the corresponding vector should be last in the array. As a user of the interface reading `initial_time`, I would expect the value to go first. My feeling is that keeping the current convention would be best.

**Review by sdrave — APPROVED** [review]

**Comment by HenKlei** [comment]
> > > I'm not sure if it's mathematically clear what 'backwards time stepping' means. Maybe one should rather write that the corresponding terminal-value problem is solved?
> > 
> > 
> > You are right. I am still not sure if the new formulation is better. And while changing the documentation, I was wondering whether one should actually change the order of the result. If one thinks of backwards time stepping the current order makes sense, however, if one sees it as a terminal value problem, the order should be reversed.
> 
> I like the 'terminal-value problem' formulation because at least it is a well-defined mathematical problem, and I think it makes more sense to speak about the problem to be solved than what a particular implementation might do.
> 
> Regarding the ordering: I guess the problem is that we use `initial_time` for the terminal value. If you think about it being the terminal value, the corresponding vector should be last in the array. As a user of the interface reading `initial_time`, I would expect the value to go first. My feeling is that keeping the current convention would be best.

Totally agreed! Thanks for clarifying this!

**Pull request merged**